test-framework/kubernetes-client/src/main/java/io/quarkus/test/kubernetes/client/PortForwardingTestResource.java

```diff
@@ -0,0 +1,128 @@
+        }
+
+        // retrieve pod name from selectors and optional index
+        final var podsResource = client().pods().inNamespace(namespace());
+        if (hasLabelSelector) {
+            podsResource.withLabelSelector(podIdentifier.labelSelector());
+        }
+        if (hasFieldSelectors) {
+            for (FieldSelector fieldSelector : podIdentifier.fieldSelectors()) {
+                switch (fieldSelector.operator()) {
+                    case eq:
+                        podsResource.withField(fieldSelector.key(), fieldSelector.value());
+                        break;
+                    case neq:
+                        podsResource.withoutField(fieldSelector.key(), fieldSelector.value());
+                    default:
+                        throw new IllegalArgumentException("Unknown field selector operator: " + fieldSelector.operator());
+                }
+            }
+        }
+        if (hasLabelValues) {
+            for (LabelValue labelValue : podIdentifier.labelValues()) {
+                final var key = labelValue.key();
+                final var value = labelValue.value();
+                if (AnnotationConstants.UNSET_STRING_VALUE.equals(value)) {
+                    podsResource.withLabel(key);
+                } else {
+                    podsResource.withLabel(key, value);
+                }
+            }
+        }
+
+        final var pods = podsResource.list().getItems();
+        final var podsNumber = pods.size();
+        final var matchingString = "matching (labels: '"
+                + podIdentifier.labelSelector() + "', fields: "
+                + Arrays.toString(podIdentifier.fieldSelectors());
+        if (podsNumber == 0) {
+            throw new IllegalArgumentException("No pod " + matchingString);
+        }
+
+        final var podIndex = podIdentifier.podIndex();
+        if (podIndex < podsNumber) {
+            pod = pods.get(podIndex);
+        } else {
+            throw new IndexOutOfBoundsException("There are only " + podsNumber
+                    + " pods " + matchingString + " but provided index was " + podIndex);
+        }
+
+        this.podName = pod.getMetadata().getName();
+
+        // deal with port
+        var portInfo = annotation.port();
+        port = portInfo.port();
+        if (AnnotationConstants.UNSET_INT_VALUE == port) {
+            port = pod.getSpec().getContainers()
+                    .get(portInfo.containerIndex())
+                    .getPorts().get(portInfo.portIndex()).getContainerPort();
+        }
+
+        // if local port is not specified, use same as container port
+        localPort = annotation.localPort();
+        if (AnnotationConstants.UNSET_INT_VALUE == localPort) {
+            localPort = port;
+        }
+    }
+
+    @Override
+    public Map<String, String> start() {
+        portForward = client().pods().inNamespace(namespace()).withName(podName).portForward(port, localPort);
```

Comment:
You can port forward, without specifying a local port:
- You don't need to worry if port is free/taken.
- User does not have to specify the port using annotations.